Kernel/Net/IPv4Socket.cpp

```diff
@@ -178,8 +178,13 @@ bool IPv4Socket::can_write(OpenFileDescription const&, u64) const
 PortAllocationResult IPv4Socket::allocate_local_port_if_needed()
 {
     MutexLocker locker(mutex());
-    if (m_local_port)
+    if (m_local_port) {
+        auto allocated = ensure_local_port_allocation();
+
+        if (allocated)
+            return { m_local_port, true };
         return { m_local_port, false };
```

Comment:
Since the boolean there is `did_allocate`, this could just be `return { m_local_port, allocated };`.

That said, should not being able to allocate a requested port be an error?

Reply:
I will have to take another look at `bind` because the port should already be allocated in the first place.

Reply:
> That said, should not being able to allocate a requested port be an error?

Doing this breaks like all of our TCP tests in `TestLibCoreStream` in different ways.

Reply:
I pushed a version that tries to combine the best from both worlds. Both `TestLibCoreStream` and manual tests with `nc` are successfull, but there might be problems in the future.